### PR TITLE
Methods for checking CUDA memory usage

### PR DESCRIPTION
**Pull request by ssnl**
1. Adds `torch.cuda.memory_cached`, `torch.cuda.max_memory_cached`, `torch.cuda.memory_allocated` and `torch.cuda.max_memory_allocated` to provide per-device memory stats. These will be useful for monitoring and benchmarking.
2. Adds two tests (single/multi-gpu) to test these four methods.

related issue: #1529

### CODE REVIEW

File: aten/src/THC/THCCachingAllocator.cpp [hunk 1]
```diff
@@ -80,6 +80,12 @@ struct THCCachingAllocator
   typedef bool (*Comparison)(const Block*, const Block*);
   typedef std::set<Block*, Comparison> FreeBlocks;
 
+  // max memory allocated
+  static size_t max_memory_allocated;
+
+  // current memory allocated
+  static size_t memory_allocated;
```
Comment (apaszke) [inline aten/src/THC/THCCachingAllocator.cpp]: It's a bit counterintuitive that this is a sum over all GPUs. One is usually interested in per-GPU usage.

File: torch/cuda/__init__.py [hunk 2]
```diff
@@ -340,6 +340,20 @@ def empty_cache():
     return torch._C._cuda_emptyCache()
 
 
+def memory_allocated():
+    """Returns the current total GPU memory usage by tensors in bytes.
+
+    .. note:: This is likely less than the amount shown in `nvidia-smi` since
+    some unused memory can be held by the cached memory allocator and some
```
Comment (apaszke) [inline torch/cuda/__init__.py]: nit: it's usually referred to as "caching allocator"

File: test/test_cuda.py [hunk 3]
```diff
@@ -425,6 +425,79 @@ def tmp(self):
...
+    def test_memory_allocated_stats(self):
+        torch.cuda.empty_cache()
+        m0 = last_m = torch.cuda.memory_allocated()
+        max_m = torch.cuda.max_memory_allocated()
+
+        def assert_change(comp=1):
+            # comp > 0: increased
+            # comp = 0: equal
+            # comp < 0: decreased
+            nonlocal last_m, max_m
```
Comment (apaszke) [inline test/test_cuda.py]: `nonlocal` is Python 3 only. You can do `last_m = [torch.cuda.memory_allocated()]` and then use `last_m[0]` everywhere

File: test/test_cuda.py [hunk 4]
```diff
@@ -425,6 +425,79 @@ def tmp(self):
...
+        assert_change(0)
+        tensors2 = permute
+        assert_change(0)
+        del permute
+        assert_change(0)
+
+        for i in range(int(N / 2)):
+            x = tensors2[i].numel()
+            del tensors2[i]
+            assert_change(-x)
```
Comment (apaszke) [inline test/test_cuda.py]: should this just be `-1`? `-x` looks as if you were trying to assert that it dropped exactly by `x` (which isn't true - you're missing a factor of `element_size`)

Comment (SsnL) [reply test/test_cuda.py]: This is used to deal with the empty tensor added at line 469 :) If the selected tensor is empty, it asserts that the memory usage stays the same.

File: aten/src/THC/THCCachingAllocator.cpp (hunk 1 above)
Comment (SsnL) [reply aten/src/THC/THCCachingAllocator.cpp]: Should the new methods instead give memory usage stats on a given GPU? I can make a map here to support that.

Comment (ezyang) [reply aten/src/THC/THCCachingAllocator.cpp]: Yes, that seems useful and strictly more general.

**Comment by ezyang** [comment]
Great, very happy to see that this happened. :)

OOC, how did you check that you hit all of the necessary entry points to modify the allocator?

**Comment by SsnL** [comment]
Echoing a comment above that is now hidden:

Should the new methods instead give memory usage stats on a given GPU? I can make a map here to support that.

**Comment by SsnL** [comment]
@ezyang I didn't do any tests, but I checked with @colesbury on that :)

**Comment by yf225** [comment]
Windows failure is the following: 

```
22:45:58 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\THC\THCCachingAllocator.cpp(172): error C2039: 'max': is not a member of 'std'
22:45:58 C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.11.25503\include\unordered_map(15): note: see declaration of 'std'
22:45:58 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\THC\THCCachingAllocator.cpp(172): error C3861: 'max': identifier not found
```

To fix it we can add `#include<algorithm>`, according to https://social.msdn.microsoft.com/Forums/vstudio/en-US/f5915ad0-a9d1-49f3-8643-ffd623f72b93/error-c2039-max-is-not-a-member-of-std?forum=vcgeneral

**Comment by SsnL** [comment]
@yf225 I'll add it, thanks!

**Comment by ezyang** [comment]
I noticed that you added stats for the "residency" (actual amount of live data), but not memory that was actually allocated from the GPU. The latter is also a pretty useful stat to collect, since (1) it is the thing that end users will actually see when the query nvidia-smi, and (2) it is the thing that will determine if you are actually out of memory or not is the actual memory allocated. Fragmentation could mean we are wasting memory in the caching allocator, that doesn't show up in the residency computation. (Though, tensor allocations tend to be big, so this isn't a big deal in practice...)

There's tons of other stats we could add to the allocator but I guess no one needed them, so we ain't gonna add 'em :)

**Comment by ezyang** [comment]
Re CI failures, I think there is an OOD. Looking.

**Comment by SsnL** [comment]
@ezyang I'd love to add actual allocated GPU usage (unused in cache + context), but I'm not sure how to count the size of the context. Perhaps just getting total cache size is fine for now?

Also, I actually think that total residency memory (i.e. memory pointed by the tensors) is sometimes a better indication for OOM because there can be large amount of available memory in cache. When nvidia-smi shows little memory left, users may still allocate large tensors.

**Comment by ezyang** [comment]
Yep, the number of blocks in the cache is the usual thing to use.

And yes, you're right, residency can be low while the cache is large, which is why residency is important too :)

**Comment by apaszke** [comment]
One way to get more detailed stats would be to just go over all blocks and accumulate information then instead of keeping some running stats. That's more expensive, but I'd assume that these functions are mostly useful for one-off debugging

File: aten/src/THC/THCCachingAllocator.cpp [hunk 5]
```diff
@@ -161,6 +167,9 @@ struct THCCachingAllocator
     allocated_blocks[block->ptr] = block;
 
     *devPtr = (void*)block->ptr;
+
+    memory_allocated += block->size;
+    max_memory_allocated = std::max(max_memory_allocated, memory_allocated);
```
Comment (peterjc123) [inline aten/src/THC/THCCachingAllocator.cpp]: Please include `<algorithm>` before using this in MSVC.

Comment (SsnL) [reply aten/src/THC/THCCachingAllocator.cpp]: will do. thanks!

**Comment by SsnL** [comment]
@apaszke Yes, but the max memory allocated is also a useful stats, so I went with the running stats route.

**Comment by ezyang** [comment]
Yeah, generally you want running stats if at all possible.

File: torch/cuda/__init__.py [hunk 6]
```diff
@@ -340,18 +340,63 @@ def empty_cache():
...
 
-def memory_allocated():
-    """Returns the current total GPU memory usage by tensors in bytes.
+def memory_allocated(device=None):
+    """Returns the current GPU memory usage by tensors in bytes for a given
+    device.
+
+    Arguments:
+        device (int): selected device. Returns statistic for the current device,
+                      given by :meth:`~torch.cuda.current_deice`, if
```
Comment (soumith) [inline torch/cuda/__init__.py]: typo here: deice -> device

File: torch/_tensor_docs.py [hunk 7]
```diff
@@ -1795,10 +1795,10 @@ def callable(a, b) -> number
                r"""
 uniform_(from=0, to=1) -> Tensor
 
-Fills :attr:`self` tensor with numbers sampled from the uniform distribution:
+Fills :attr:`self` tensor with numbers sampled from the continuous uniform
+distribution:
 
 .. math:
-
     P(x) = \dfrac{1}{to - from}
```
Comment (SsnL) [inline torch/_tensor_docs.py]: remove the empty line so it shows up in html.

File: torch/cuda/__init__.py (hunk 6 above)
Comment (SsnL) [reply torch/cuda/__init__.py]: updated :)

**Comment by SsnL** [comment]
Added cache stats. Changed the functions to provide per-device stats.

**Comment by SsnL** [comment]
The build errors are fixed in #4544

File: docs/source/notes/cuda.rst [hunk 8]
```diff
@@ -48,8 +48,12 @@ Memory management
 PyTorch use a caching memory allocator to speed up memory allocations. This
 allows fast memory deallocation without device synchronizations. However, the
 unused memory managed by the allocator will still show as if used in
-`nvidia-smi`. Calling :meth:`~torch.cuda.empty_cache` can release all unused
-cached memory from PyTorch so that those can be used by other GPU applications.
+`nvidia-smi`. You can use :meth:`~torch.cuda.memory_allocated` and
+:meth:`~torch.cuda.max_memory_allocated` to monitor memory occupied by
+tensors, and use :meth:`~torch.cuda.memory_cached` and
+:meth:`~torch.cuda.max_memory_cachted` to monitor memory managed by the caching
```
Comment (jlquinn) [inline docs/source/notes/cuda.rst]: misspelling - should be torch.cuda.max_memory_cached

Comment (SsnL) [reply docs/source/notes/cuda.rst]: thanks, fixed!

File: test/test_cuda.py [hunk 9]
```diff
@@ -425,6 +425,167 @@ def tmp(self):
...
+        m0 = torch.cuda.memory_allocated(device)
+        last_m_arr = [torch.cuda.memory_allocated(device)]
+        max_m_arr = [torch.cuda.max_memory_allocated(device)]
+        last_c_arr = [torch.cuda.memory_cached(device)]
+        max_c_arr = [torch.cuda.max_memory_cached(device)]
+
+        def alloc(*size):
+            with torch.cuda.device(device):
+                # NOTE: do **not** use methods that will have additional
+                #       overhead, e.g., inplace random sampling methods.
```
Comment (apaszke) [inline test/test_cuda.py]: why?

File: test/test_cuda.py [hunk 10]
```diff
@@ -425,6 +425,167 @@ def tmp(self):
...
+        assert_change(-1)
+        assert_change(0)
+        self.assertEqual(torch.cuda.memory_allocated(device), m1)
+        yield True
+
+        del tensors1
+        assert_change(-1)
+        self.assertEqual(torch.cuda.memory_allocated(device), m0)
+
+        assert_change(0, empty_cache=True)
```
Comment (apaszke) [inline test/test_cuda.py]: Why do you empty cache here? Just to test it?

**Review by apaszke — APPROVED** [review]

File: test/test_cuda.py (hunk 9 above)
Comment (SsnL) [reply test/test_cuda.py]: First time calling inplace cuda sampling methods will alloc an RNG state. After del the tensor, the state still remains in GPU memory, causing equal checks at line 538 to fail.

File: test/test_cuda.py (hunk 10 above)
Comment (SsnL) [reply test/test_cuda.py]: Yes, just to test it.

**Comment by SsnL** [comment]
again test fails are unrelated, and fixed in #4544

**Pull request merged**